File: Basics/MinStack.py
class Minstack:
	"""docstring for Minstack"""
	def __init__(self):
		self.stack = []
		self.n = 0
		self.mini = -1

	def getMin(self):
		if self.n == 0:
			return -1
		return self.mini

	def push(self,x):
		if self.n==0:
			self.stack.append(x)
			self.mini = x
		else:
			if x<=self.getMin():
				self.stack.append(2*x-self.getMin())
				self.mini = x
			else:
				self.stack.append(x)
		self.n+=1
		return

	def pop(self):
		if self.n==0:
			return
		tmp = self.stack.pop()
		mn = self.getMin()
		if tmp<=mn:
			self.mini = 2*mn - tmp
		self.n-=1
		return

	def top(self):
		if self.n==0:
			return -1
		tmp = self.stack[self.n-1]
		mn = self.getMin()
		if tmp<=mn:
			return mn
		return tmp

if __name__ == '__main__':
	stack = Minstack()
	stack.push(10)
	stack.push(24)
	stack.push(7)
	stack.push(3)
	stack.push(9)
	stack.push(1)
	stack.push(4)
	stack.push(8)
	print(stack.top(),stack.getMin())
	stack.pop()
	print(stack.top(),stack.getMin())
	stack.pop()
	print(stack.top(),stack.getMin())
	stack.pop()
	print(stack.top(),stack.getMin())
	stack.pop()
	print(stack.top(),stack.getMin())
	stack.pop()
	print(stack.top(),stack.getMin())

File: Interview_Bit/Stacks_and_Queues/Min_stack.cpp
stack<int> min_stack;
int minElem = INT_MAX;
MinStack::MinStack() {
    while(!min_stack.empty())
        min_stack.pop();
    minElem = INT_MAX;
}
void MinStack::push(int x) {
    if(min_stack.empty()){
        min_stack.push(x);
        minElem = x;
    }
    else{
        if(x>=minElem)
            min_stack.push(x);
        else{
            min_stack.push(2*x - minElem);
            minElem = x;
        }
    }
}

void MinStack::pop() {
    if(!min_stack.empty()){
        int y = min_stack.top();
        min_stack.pop();
        if(y<minElem){
            minElem = 2*minElem - y;
        }
    }
}

int MinStack::top() {
    if(min_stack.empty())
        return -1;
    int t = min_stack.top();
    return (t<minElem) ? minElem : t;
}

int MinStack::getMin() {
    if(min_stack.empty())
        return -1;
    return minElem;
}

